docs.md

```diff
@@ -703,6 +703,14 @@ derive_tagged_enum_destructor = false
 # default: false
 derive_tagged_enum_copy_constructor = false
 
+# Whether enums with fields should generate an empty, private destructor.
+# This allows the auto-generated constructor functions to compile, if there are
+# non-trivially constructible members. This falls in the same family of
+# dangerousness as `derive_tagged_enum_copy_constructor` and co.
```

Comment:
i'm completely unfamiliar with the issue and having trouble making sense of it, can you expand on this? what types don't we support? i'm not familiar with the concept of "trivially constructible".

Reply:
So the issue is the following. Simplified.

This builds, as expected:

```c++
#include <cstdint>
#include <new>

struct TypeWithDestructor {
    uint8_t field;
    // ~TypeWithDestructor() {}
};

struct TaggedUnion {
    static TaggedUnion Create(const TypeWithDestructor& a) {
        TaggedUnion ret;
        ret.type = 0;
        new (&ret.foo) TypeWithDestructor(a);
        return ret;
    }

    uint8_t type;
    union {
        TypeWithDestructor foo;
    };
};
```

If you uncomment the destructor, then it [doesn't](https://godbolt.org/z/NekDlQ), because C++ will implicitly delete the default constructor of `TaggedUnion`.

Since all we want is a default constructor that doesn't initialize anything, so that `Create()` does the right thing, a private default constructor serves this purpose. It also avoids people accidentally creating uninitialized structs themselves, which is a win-win.

Reply:
The final simplified example for cbindgen would be:

```c++
#include <cstdint>
#include <new>

struct TypeWithDestructor {
    uint8_t field;
    ~TypeWithDestructor() {}
};

struct TaggedUnion {
    static TaggedUnion Create(const TypeWithDestructor& a) {
        TaggedUnion ret;
        ret.type = 0;
        new (&ret.foo) TypeWithDestructor(a);
        return ret;
    }

    private:
    TaggedUnion() {}
    
    public:
    ~TaggedUnion() { /* cbindgen-generated */ }

    uint8_t type;
    union {
        TypeWithDestructor foo;
    };
};
```

With nicer enum types and such of course.

Reply:
Is the claim that this is dangerous just incorrect? Adding a private default constructor doesn't actually change the ABI in any meaningful way, does it? It just enables you to add destructors that *do* mess up the ABI..?

Reply:
Adding constructors changes the ABI of return types in MSVC: https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019#return-values

> To return a user-defined type by value in RAX, it must have a length of 1, 2, 4, 8, 16, 32, or 64 bits. It must also have no user-defined constructor, destructor, or copy assignment operator; no private or protected non-static data members; no non-static data members of reference type; no base classes; no virtual functions; and no data members that do not also meet these requirements. (This is essentially the definition of a C++03 POD type. Because the definition has changed in the C++11 standard, we don't recommend using std::is_pod for this test.)

Reply:
And yeah I wish I didn't know this :)